swarm/network/pss.go

```diff
@@ -1,52 +1,384 @@
+	//PeerReverse map[adapters.NodeId]pot.Address // keep track of all virtual p2p.Peers we are currently speaking to
+	handlers map[PssTopic]func([]byte, *p2p.Peer, []byte) error // topic and version based pss payload handlers
+	fwdcache map[pssDigest]time.Time                            // checksum of unique fields from pssmsg mapped to expiry, cache to determine whether to drop msg
+	cachettl time.Duration                                      // how long to keep messages in fwdcache
+	hasher   hash.Hash                                          // hasher to digest message to cache
 }
 
-func NewPss(k Overlay, addr []byte) *Pss {
+// 32bit checks should be enough for this use
+func (self *Pss) hashMsg(msg *PssMsg) pssDigest {
+	self.hasher.Reset()
+	self.hasher.Write(msg.To)
+	self.hasher.Write(msg.Payload.SenderUAddr)
+	self.hasher.Write(msg.Payload.SenderOAddr)
+	self.hasher.Write(msg.Payload.Topic[:])
+	self.hasher.Write(msg.Payload.Payload)
+	b := self.hasher.Sum([]byte{})
+	return pssDigest(binary.BigEndian.Uint32(b))
+}
+
+// todo error check overlay integrity
+func NewPss(k Overlay, cachettl time.Duration) *Pss {
+	if cachettl == 0 {
+		cachettl = defaultDigestCacheTTL
+	}
 	return &Pss{
-		Overlay: k,
-		LocalAddr: addr,
-		C: make(chan []byte),
+		Overlay:  k,
+		NodeId:   adapters.NewNodeId(k.GetAddr().UnderlayAddr()),
+		PeerPool: make(map[pot.Address]map[PssTopic]*PssReadWriter, PssPeerCapacity),
+		//PeerReverse: make(map[adapters.NodeId]pot.Address, PssPeerCapacity),
+		handlers: make(map[PssTopic]func([]byte, *p2p.Peer, []byte) error),
+		fwdcache: make(map[pssDigest]time.Time),
+		cachettl: cachettl,
+		hasher:   storage.MakeHashFunc("SHA3")(),
 	}
 }
 
-type PssMsg struct {
-	To   []byte
-	Data	[]byte
+// when full oldest out
+// periodically clean cache
+func (self *Pss) addToFwdCache(digest pssDigest) error {
+	self.fwdcache[digest] = time.Now().Add(self.cachettl)
+	return nil
 }
 
-func (pm *PssMsg) String() string {
-	return fmt.Sprintf("PssMsg: Recipient: %v", pm.To)
+// returns true if in cache AND not expired
+func (self *Pss) checkCache(digest pssDigest) bool {
+	entry, ok := self.fwdcache[digest]
+	if !ok || entry.Before(time.Now()) {
+		return false
+	}
+	return true
+}
+
+func (self *Pss) Register(name string, version int, handler func(msg []byte, p *p2p.Peer, from []byte) error) error {
+	t, err := MakeTopic(name, version)
+	if err != nil {
+		return err
+	}
+	self.handlers[t] = handler
+	return nil
+}
+
+func (self *Pss) GetHandler(topic PssTopic) func([]byte, *p2p.Peer, []byte) error {
+	return self.handlers[topic]
 }
 
-func (ps *Pss) HandlePssMsg(msg interface{}) error {
-	pssmsg := msg.(*PssMsg)
-	to := pssmsg.To
-	if bytes.Equal(to, ps.LocalAddr) {
-		glog.V(logger.Detail).Infof("Pss to us, yay!", to)
-		ps.C <- pssmsg.Data
+func (self *Pss) AddPeerTopic(id pot.Address, topic PssTopic, rw *PssReadWriter) error {
```

Comment:
this should not be exported. 
Instead there should be an AddPeer call on PssProtocol as part of the API